### PR TITLE
Document using httpx for async tests

### LINKED ISSUE #1717: Document how to use `httpx.AsyncClient` as test client for async setup
I'd like for us to document how to use `httpx.AsyncClient` as test client on the [testclient](https://www.starlette.io/testclient/) page, or at least add a note about it with a reference to [`httpx` section "Calling into Python Web Apps"](https://www.python-httpx.org/async/#calling-into-python-web-apps).

The idea here is just to present the possibility to use the "async test setup".

Note: this issue is not in conflict with any other discussion about removing the `TestClient` from Starlette or replacing `requests` by `httpx` internally.

### PR DESCRIPTION
**Pull request by adriangb**
Closes #1717

### CODE REVIEW

File: docs/testclient.md [hunk 1]
```diff
@@ -154,3 +154,40 @@ May raise `starlette.websockets.WebSocketDisconnect`.
...
+def hello(request: Request) -> PlainTextResponse:
+    return PlainTextResponse("Hello World!")
+
+
+app = Starlette(routes=[Route("/", hello)])
+
+
+# if you're using pytest, you'll need to to add an async marker like:
+# @pytest.mark.anyio  # using https://github.com/agronholm/anyio
+# @pytest.mark.asyncio  # using https://github.com/pytest-dev/pytest-asyncio
```
Comment (euri10) [inline docs/testclient.md]: it's not needed for pytest-asyncio, see https://github.com/pytest-dev/pytest-asyncio#auto-mode
> When the mode is auto, all discovered async tests are considered asyncio-driven even if they have no @pytest.mark.asyncio marker.

for pytest-anyio I don't know
```suggestion
```

File: docs/testclient.md [hunk 2]
```diff
@@ -154,3 +154,40 @@ May raise `starlette.websockets.WebSocketDisconnect`.
...
+async def test_app() -> None:
+    async with AsyncClient(app=app, base_url="http://testserver") as client:
```
Comment (euri10) [inline docs/testclient.md:188]: I'd insist on having the `base_url` set, it's important, and if not passed correctly the error in httpx is kind of useless, see https://github.com/encode/httpx/issues/1438

**Review by euri10 — CHANGES_REQUESTED** [review]

File: docs/testclient.md (hunk 1 above)
Comment (adriangb) [reply docs/testclient.md]: TIL 😄

File: docs/testclient.md [hunk 3]
```diff
@@ -154,3 +154,39 @@ May raise `starlette.websockets.WebSocketDisconnect`.
...
+
+def hello(request: Request) -> PlainTextResponse:
+    return PlainTextResponse("Hello World!")
+
+
+app = Starlette(routes=[Route("/", hello)])
+
+
+# if you're using pytest, you'll need to to add an async marker like:
+# @pytest.mark.anyio  # using https://github.com/agronholm/anyio
```
Comment (adriangb) [inline docs/testclient.md:184]:
```suggestion
# @pytest.mark.anyio  # using https://github.com/agronholm/anyio
# or install and configure pytest-asyncio (https://github.com/pytest-dev/pytest-asyncio)
```

File: docs/testclient.md [hunk 4]
```diff
@@ -154,3 +154,39 @@ May raise `starlette.websockets.WebSocketDisconnect`.
...
+async def test_app() -> None:
+    async with AsyncClient(app=app, base_url="http://testserver") as client:
```
Comment (adriangb) [inline docs/testclient.md:188]:
```suggestion
    # note: you _must_ set `base_url` for relative urls like "/" to work
    async with AsyncClient(app=app, base_url="http://testserver") as client:
```

File: docs/testclient.md (hunk 2 above)
Comment (adriangb) [reply docs/testclient.md:188]: added a comment

File: docs/testclient.md (hunk 1 above)
Comment (adriangb) [reply docs/testclient.md]: Does https://github.com/encode/starlette/pull/1728#discussion_r915033286 sound good?

File: docs/testclient.md (hunk 3 above)
Comment (euri10) [reply docs/testclient.md:184]: yup that works too :)

**Review by euri10 — APPROVED** [review]
lgtm, thanks a lot @adriangb :)

File: docs/testclient.md [hunk 5]
```diff
@@ -154,3 +154,41 @@ May raise `starlette.websockets.WebSocketDisconnect`.
...
+def hello(request: Request) -> PlainTextResponse:
+    return PlainTextResponse("Hello World!")
+
+
+app = Starlette(routes=[Route("/", hello)])
+
+
+# if you're using pytest, you'll need to to add an async marker like:
+# @pytest.mark.anyio  # using https://github.com/agronholm/anyio
+# or install and configure pytest-asyncio (https://github.com/pytest-dev/pytest-asyncio)
```
Comment (Kludex) [inline docs/testclient.md:185]: Can we add this as a note below instead?

Comment (adriangb) [reply docs/testclient.md:185]: Sure I'll make a followup

Comment (Kludex) [reply docs/testclient.md:185]: Thanks ❤️

**Pull request merged**